include/cpuid.h

```diff
@@ -32,6 +32,83 @@
 #define CPUID_BRAND_INFO_MIN 0x80000002U
 #define CPUID_BRAND_INFO_MAX 0x80000004U
 
+#define MHZ_MULTIPLIER 1000000
+#define GHZ_MULTIPLIER 1000000000
+
+static inline uint64_t get_cpu_freq(const char *cpu_str) {
+    size_t len = strlen(cpu_str);
+    uint64_t multiplier = 0;
+    uint64_t frequency = 0;
+    char buf[16];
+    char buf2[16];
+    char *reverse = &buf[0];
+    char *freq = &buf2[0];
+
+    /* we need to reverse the vendor string for parsing the freq */
+    while (len--) {
+        if (isspace(*(cpu_str + len))) {
+            *(reverse) = '\0';
+            break;
+        }
+        *(reverse++) = *(cpu_str + len);
+    }
+
+    if (!strstr(buf, "zHM") && !strstr(buf, "zHG")) {
+        // if we have nothing to calculate, just return
+        return 0;
+    }
+
+    if (strstr(buf, "zHM")) {
+        multiplier = MHZ_MULTIPLIER;
+
+        for (int i = strlen(buf) - 1; i >= 0; i--)
+            if (isdigit(buf[i]))
+                *(freq++) = buf[i];
+
+        frequency = strtoul(buf2, NULL, 0) * multiplier;
+    }
+    else if (strstr(buf, "zHG")) {
+        multiplier = GHZ_MULTIPLIER;
+
+        /* Convert a floating number to
+         * n * GHz + m * MHz
+         */
+        for (int i = strlen(buf) - 1; i >= 0; i--) {
+            if (isdigit(buf[i]))
+                *(freq++) = buf[i];
+            if (ispunct(buf[i])) {
+                *(freq) = '\0';
+                frequency = strtoul(buf2, NULL, 0) * multiplier;
+                memset(buf2, 0, sizeof(buf2));
+                freq = &buf2[0];
+                multiplier = MHZ_MULTIPLIER;
+            }
+        }
+        *(freq) = '\0';
+
+        /* we have to check if we hit a float and calculate
+         * the length of the MHz portion e.g. 2.80GHz or 2.8GHz
+         */
+        if (multiplier == MHZ_MULTIPLIER) {
+            if (strlen(buf2) == 2)
+                frequency += strtoul(buf2, NULL, 0) * 10 * multiplier;
+            else if (strlen(buf2) == 1)
+                frequency += strtoul(buf2, NULL, 0) * 100 * multiplier;
```

Comment:
this could be: `frequency += strtoul(buf2, NULL, 0) * (1000 / (10 * strlen(buf2))) * multiplier`